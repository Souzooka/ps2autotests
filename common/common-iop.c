#include <stdio.h>
#include <stdarg.h>
#include <xprintf.h>

#define SCHEDF_BUFFER_SIZE 65536

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsnprintf(schedfBuffer + schedfBufferPos, SCHEDF_BUFFER_SIZE, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}
