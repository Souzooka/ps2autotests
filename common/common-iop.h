#ifndef PS2AUTOTESTS_COMMON_IOP_H
#define PS2AUTOTESTS_COMMON_IOP_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _IOP
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

void schedf(const char *format, ...);
void flushschedf();

#endif
