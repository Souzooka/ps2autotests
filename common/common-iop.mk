# _____     ___ ____     ___ ____
#  ____|   |    ____|   |        | |____|
# |     ___|   |____ ___|    ____| |    \    PS2DEV Open Source Project.
#-----------------------------------------------------------------------
# Copyright 2001-2004, ps2dev - http://www.ps2dev.org
# Licenced under Academic Free License version 2.0
# Review ps2sdk README & LICENSE files for further details.
#
# $Id: Makefile.eeglobal_cpp_sample 1525 2009-01-23 00:24:13Z radad $

# Define pefixes for the various toolchain binaries.
include $(COMMON_DIR)/common-defines.mk

ASFLAGS_TARGET = -mcpu=r3000

CFLAGS_TARGET  = -miop
ASFLAGS_TARGET = -march=r3000
LDFLAGS_TARGET = -miop

# include dir
IOP_INCS := -I$(PS2SDK)/iop/include -I$(PS2SDK)/common/include -I$(COMMON_DIR) \
	-I. $(IOP_INCS)
# C compiler flags
IOP_CFLAGS := $(CFLAGS_TARGET) -O2 -G0 $(IOP_INCS) $(IOP_CFLAGS)

# linker flags
IOP_LDFLAGS := $(LDFLAGS_TARGET) -nostdlib -L$(PS2SDK)/iop/lib $(IOP_LDFLAGS)

# asssembler flags
IOP_ASFLAGS := $(ASFLAGS_TARGET) -EL -G0 $(IOP_ASFLAGS)

# link with following libraries (libs need to be defined multiple times in order for linking to work!!)
IOP_LIBS += -lkernel -lgcc

# Externally defined variables: IOP_BIN, IOP_OBJS, IOP_LIB

%.o : %.c
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.S
	$(IOP_CC) $(IOP_CFLAGS) -c $< -o $@

%.o : %.s
	$(IOP_AS) $(IOP_ASFLAGS) $< -o $@

%.o : %.lst
	echo "#include \"irx_imports.h\"" > build-imports.c
	cat $< >> build-imports.c
	$(IOP_CC) $(IOP_CFLAGS) -c build-imports.c -o $@
	rm -f build-imports.c

%.irx : %.o $(EXTRA_OBJS)
	$(IOP_CC) $(IOP_LDFLAGS) -o $@ $< $(EXTRA_OBJS) $(IOP_LIBS)

all: $(TARGETS:=.irx)

clean:
	-$(RM) -f $(TARGETS:=.irx) $(TARGETS:=.o)

rebuild: clean all
