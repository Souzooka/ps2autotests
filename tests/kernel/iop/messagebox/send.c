#include <common-iop.h>
#include <thbase.h>
#include <thmsgbx.h>

#define STACK_SIZE 0x800

typedef struct
{
	iop_message_t header;
	u32           payload;
} MSG;

u32 g_receivedPayload = 0;

void messageWaitThreadProc(u32 mbxId) {
	MSG* msg = NULL;
	ReceiveMbx((void**)&msg, mbxId);
	g_receivedPayload = msg->payload;
}

void printMbx(s32 mbxId) {
	iop_mbx_status_t mbxInfo;
	s32 result = ReferMbxStatus(mbxId, &mbxInfo);
	if(result >= 0) {
		MSG* message = (MSG*)mbxInfo.topPacket;
		u32 i = 0;
		printf("wait threads: %d, messages: %d", mbxInfo.numWaitThreads, mbxInfo.numMessage);
		if(mbxInfo.numMessage != 0) {
			printf(" -> ");
		}
		for(i = 0; i < mbxInfo.numMessage; i++) {
			printf("{ payload: %d, priority: %d }", message->payload, message->header.priority);
			if((i + 1) != mbxInfo.numMessage) {
				printf(", ");
			}
			message = (MSG*)message->header.next;
		}
		printf("\n");
	} else {
		printf("  failed (%d)\n", result);
	}
}

int getThreadPriority(int threadId) {
	iop_thread_info_t threadInfo = {};
	ReferThreadStatus(threadId, &threadInfo);
	return threadInfo.currentPriority;
}

void testMultipleSend(u32 mbxAttr) {
	static const int msgCount = 3;
	int i = 0;
	MSG msg[msgCount];
	
	iop_mbx_t mbxInfo = {};
	mbxInfo.option = 0;
	mbxInfo.attr   = mbxAttr;
	s32 mbxId = CreateMbx(&mbxInfo);

	for(i = 0; i < msgCount; i++) {
		memset(&msg[i], 0, sizeof(MSG));
		msg[i].header.priority = msgCount - i;
		msg[i].payload         = i + 1;
		SendMbx(mbxId, &msg[i]);
	}

	printf("  box contents: ");
	printMbx(mbxId);
	
	DeleteMbx(mbxId);
}

void testSendWakeup() {
	iop_mbx_t mbxInfo = {};
	mbxInfo.option = 0;
	mbxInfo.attr   = MBA_THFIFO | MBA_MSFIFO;
	s32 mbxId = CreateMbx(&mbxInfo);

	iop_thread_t threadInfo = {};
	threadInfo.attr      = TH_C;
	threadInfo.thread    = (void*)&messageWaitThreadProc;
	threadInfo.priority  = getThreadPriority(TH_SELF) - 1;
	threadInfo.stacksize = STACK_SIZE;
	s32 thId = CreateThread(&threadInfo);
	StartThread(thId, (void*)mbxId);
	
	printf("  box contents after thread creation: ");
	printMbx(mbxId);

	MSG msg = {};
	msg.payload = 0xBEEF;
	SendMbx(mbxId, &msg);
	
	printf("  box contents after send: ");
	printMbx(mbxId);
	
	printf("  received payload: 0x%X\n", g_receivedPayload);
}

int _start(int argc, char **argv) {
	printf("-- TEST BEGIN\n");
	
	printf("sending (fifo message queuing):\n");
	testMultipleSend(MBA_THFIFO | MBA_MSFIFO);
	
	printf("sending (prio message queuing):\n");
	testMultipleSend(MBA_THFIFO | MBA_MSPRI);
	
	printf("sending wakeup behavior:\n");
	testSendWakeup();
	
	printf("invalid params:\n");
	{
		int result = SendMbx(~0, NULL);
		printf("  sending message to invalid box: %d\n", result);

		//Sending a null message does seem to work even though it probably has
		//dangerous side effects since the message header at address 0 will be modified
		//int result = SendMbx(mbxId, NULL);
		//printf("sending null message: %d\n", result);
		//printMbx(mbxId);
	}
	
	printf("-- TEST END\n");
	return 0;
}
