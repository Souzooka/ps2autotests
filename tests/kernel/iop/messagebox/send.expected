-- TEST BEGIN
sending (fifo message queuing):
  box contents: wait threads: 0, messages: 3 -> { payload: 1, priority: 3 }, { payload: 2, priority: 2 }, { payload: 3, priority: 1 }
sending (prio message queuing):
  box contents: wait threads: 0, messages: 3 -> { payload: 3, priority: 1 }, { payload: 2, priority: 2 }, { payload: 1, priority: 3 }
sending wakeup behavior:
  box contents after thread creation: wait threads: 1, messages: 0
  box contents after send: wait threads: 0, messages: 0
  received payload: 0xBEEF
invalid params:
  sending message to invalid box: -410
-- TEST END
